python_modules/dagster/dagster/_core/op_concurrency_limits_counter.py

```diff
@@ -0,0 +1,140 @@
+import json
+from collections import defaultdict
+from typing import Mapping, Optional, Sequence
+
+import pendulum
+
+from .instance import DagsterInstance
+from .snap import JobSnapshot
+from .storage.dagster_run import DagsterRun, DagsterRunStatus, RunRecord
+from .storage.tags import (
+    GLOBAL_CONCURRENCY_TAG,
+    RUN_OP_ROOT_CONCURRENCY_KEYS,
+)
+
+
+def compute_concurrency_tags_for_snapshot(job_snapshot: JobSnapshot) -> Mapping[str, str]:
+    """Utility function called at run creation time to tag each run with the concurrency info needed
+    to keep track of concurrency limits for each in-flight run.
+    """
+    concurrency_keys_by_node_name = {
+        node.node_name: node.tags.get(GLOBAL_CONCURRENCY_TAG)
+        for node in job_snapshot.dep_structure_snapshot.node_invocation_snaps
+        if node.tags.get(GLOBAL_CONCURRENCY_TAG)
+    }
+    if not concurrency_keys_by_node_name:
+        return {}
+    root_nodes = [
+        node.node_name
+        for node in job_snapshot.dep_structure_snapshot.node_invocation_snaps
+        if not node.input_dep_snaps
```

Comment:
actually, this doesn't seem to be a reliable way of calculating the root nodes for a job snapshot... will amend